cosi/cosi.go

```diff
@@ -1,450 +1,399 @@
+// it is present in the list of keys and sets the corresponding index in the
+// bitmask to 1 (enabled).
+func NewMask(suite abstract.Suite, publics []abstract.Point, myKey abstract.Point) (*Mask, error) {
+	m := &Mask{
+		publics: publics,
+	}
+	m.mask = make([]byte, m.Len())
+	m.AggregatePublic = suite.Point().Null()
+	if myKey != nil {
+		found := false
+		for i, key := range publics {
+			if key.Equal(myKey) {
+				m.SetBit(i, true)
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, errors.New("key not found")
+		}
+	}
+	return m, nil
 }
 
-// GetCommitment returns the commitment generated by this CoSi (not aggregated).
-func (c *CoSi) GetCommitment() abstract.Point {
-	return c.commitment
+// Mask returns a copy of the participation bitmask.
+func (m *Mask) Mask() []byte {
+	clone := make([]byte, len(m.mask))
+	copy(clone[:], m.mask)
+	return clone
 }
 
-// GetResponse returns the individual response generated by this CoSi
-func (c *CoSi) GetResponse() abstract.Scalar {
-	return c.response
+// SetMask sets the participation bitmask according to the given byte slice
+// interpreted in little-endian order, i.e., bits 0-7 of byte 0 correspond to
+// cosigners 0-7, bits 0-7 of byte 1 correspond to cosigners 8-15, etc.
+func (m *Mask) SetMask(mask []byte) error {
+	if m.Len() != len(mask) {
+		return fmt.Errorf("mismatching mask lengths")
+	}
+	for i := range m.publics {
+		byt := i >> 3
+		msk := byte(1) << uint(i&7)
+		if ((m.mask[byt] & msk) == 0) && ((mask[byt] & msk) != 0) {
+			m.mask[byt] ^= msk // flip bit in mask from 0 to 1
+			m.AggregatePublic.Add(m.AggregatePublic, m.publics[i])
+		}
+		if ((m.mask[byt] & msk) != 0) && ((mask[byt] & msk) == 0) {
+			m.mask[byt] ^= msk // flip bit in mask from 1 to 0
+			m.AggregatePublic.Sub(m.AggregatePublic, m.publics[i])
+		}
+	}
+	return nil
 }
 
-// genCommit generates a random scalar vi and computes its individual commit
-// Vi = G^vi
-func (c *CoSi) genCommit(s cipher.Stream) {
-	var stream = s
-	if s == nil {
-		stream = random.Stream
-	}
-	c.random = c.suite.Scalar().Pick(stream)
-	c.commitment = c.suite.Point().Mul(nil, c.random)
-	c.aggregateCommitment = c.commitment
+// Len returns the mask length in bytes.
+func (m *Mask) Len() int {
+	return (len(m.publics) + 7) >> 3
 }
 
-// genResponse creates the response
-func (c *CoSi) genResponse() error {
-	if c.private == nil {
-		return errors.New("No private key given in this cosi")
+// SetBit enables (enable: true) or disables (enable: false) the bit
+// in the participation mask of the given cosigner.
+func (m *Mask) SetBit(signer int, enable bool) error {
+	if signer > len(m.publics) {
+		return errors.New("index out of range")
 	}
-	if c.random == nil {
-		return errors.New("No random scalar computed in this cosi")
+	byt := signer >> 3
+	msk := byte(1) << uint(signer&7)
+	if ((m.mask[byt] & msk) == 0) && enable {
+		m.mask[byt] ^= msk // flip bit in mask from 0 to 1
```

Comment:
methinks a `|=` for setting and a `&^=` for clearing would be more understandable. Or even define a `BitClear` and a `BitSet` that does the `AggregatePublic.Sub` and `AggregatePublic.Add`, too, as it is used both here and in `SetMask`.